Add docs page on deriving array rules (#191)
JuliaDiff/ChainRulesCore.jl: docs/src/arrays.md

# Arrays

When the inputs and outputs of the functions are arrays (potentially with scalars), a modified version of Giles' method is a quick way of deriving `frule`s and `rrule`s.

> Giles' method is succinctly explained in [this paper](https://people.maths.ox.ac.uk/gilesm/files/NA-08-01.pdf), but we will slightly generalize it.

Reviewer: Perhaps comment on why it's been generalised, otherwise the reader won't discover it until they read your `!!!note` further down.

Author: Good idea, I'll introduce the comparison of the note at the top.

$$dC = \frac{\partial f}{\partial A} dA + \frac{\partial f}{\partial B} dB.$$

The pushforward is obtained by dividing by the differential wrt a single scalar input $t$.
Writing $\dot{C} = \frac{dC}{dt}$, we replace the differentials with the forward mode sensitivities:

$$\dot{C} = \frac{\partial f}{\partial A} \dot{A} + \frac{\partial f}{\partial B} \dot{B}.$$

The terms $\frac{\partial f}{\partial A}$ are array-array derivatives, i.e. a type of Jacobian.
We do not write these down explicitly, but we instead use differential identities to write down the terms $\frac{\partial f}{\partial A} dA$, which as we've seen behave like the Jacobian-vector-products $\frac{\partial f}{\partial A} \dot{A}$.
The differential identities follow directly from the usual scalar identities.

### Matrix addition

$$C = A + B$$

$$dC = dA + dB$$

### Matrix multiplication

$$C = A B$$

First we write in component form.

$$C_{ij} = \sum_k A_{ik} B_{kj}$$

Then we apply the product rule to get the scalar differential identity.

$$dC_{ij} = \sum_k dA_{ik}\ B_{kj} + A_{ik}\ dB_{kj}$$

But this is just a sum matrix products:

> $$C = dA\ B + A\ dB$$

Reviewer 2: ```suggestion
$$dC = dA\ B + A\ dB$$
```

So we now have the matrix product rule.

### Matrix inversion

> $$C A = A C = I$$

Reviewer: For the sake of flow, maybe start by stating `C := inv(A)` or whatever to make it clear what function the rule is being derived for. When you're reading these derivations for the first time I think it's helpful to make it clear what the input is and what the output is so that you don't get lost.

Reviewer 2: What do you think about just using `A^{-1}` instead of `C`?

Here we differentiate the constraint, applying the matrix product rule, and solve for $dC$.

$$dC\ A + C\ dA = 0$$
$$dC\ A\ C + C\ dA\ C = 0$$

> $$dC = -C\ dA\ C$$

Reviewer: I think you meant to invert both occurrences of `C` in this line.

Author: Oh, the slashes actually aren't characters, they add more spacing between the characters, since otherwise it's tricky to see that `dA` is one unit. Perhaps there's a way to do that that is friendlier to read the source of.

Reviewer 3: Perhaps `~` which is nonbreaking space, or `\;`,  `\quad` etc?

### Other useful identities

Two useful identities are $d(A^H) = dA^H$ and $d(A^T) = dA^T$, where $\cdot^H$ is the conjugate transpose.

## Writing reverse-mode rules

Reverse-mode rules are a little more involved.
For a real scalar function $s = g(C)$, the differential of $s$ is the real part of the inner product of the adjoint of $C$, $\overline{C}$, and the differential of $C$:

$$ds = \Re\langle \overline{C}, dC \rangle = \Re\left(\sum_{i,\dots,j} \overline{C}_{i,\dots,j}^* dC_{i,\dots,j}\right),$$

where $\cdot^*$ is the complex conjugate, and $\Re(x)$ is the real part of $x$ (i.e. `real(x)`).

For matrices and vectors, we can write this as

$$ds = \Re(\operatorname{tr}(\overline{C}^H\ dC)),$$

where $\operatorname{tr}$ is the `trace`.

> Plugging in the identity for $dC$, we get

Reviewer: ```suggestion
Plugging in the expression for $dC$, we get
```
On first reading I thought you meant the identity _matrix_.

$$ds = \Re\left(\operatorname{tr}\left(\overline{C}^H \frac{\partial f}{\partial A} dA + \overline{C}^H\frac{\partial f}{\partial B} dB\right)\right) = \Re\left(\operatorname{tr}\left(\overline{C}^H \frac{\partial f}{\partial A} dA\right)\right) + \Re\left(\operatorname{tr}\left(\overline{C}^H\frac{\partial f}{\partial B} dB\right)\right)$$

By applying the same definition of $ds$ to the intermediates $A$ and $B$, we get

> $$ds = \Re\left(\operatorname{tr}(\overline{A}^H\ dA)\right) + \Re\left(\operatorname{tr}(\overline{B}^H\ dB)\right)$$

Reviewer: I think another line here that explicitly highlights the fact that you've equated `\overline{A}^H` and ``\overline{C}^H df / dA` would be helpful. There's a lot of notation flying around here, so it might well take someone who's new to this a little while to figure out that you're just equating terms.

Subsequently, manipulate into this form and solve for the adjoint derivatives of the inputs.
Several properties of `trace` make this easier:

$$\operatorname{tr}(A+B) = \operatorname{tr}(A) + \operatorname{tr}(B)$$
$$\operatorname{tr}(A^T) = \operatorname{tr}(A)$$
$$\operatorname{tr}(A^H) = \operatorname{tr}(A)^*$$
$$\operatorname{tr}(AB) = \operatorname{tr}(BA)$$

!!! note
    Our method is identical to Giles' approach, except we have replace the transpose with the conjugate transpose, and we have added the constraint that the inner product be real.
    This produces the correct pullbacks for arrays with complex entries.

Here are a few examples.

### Matrix multiplication

We above derived

$$dC = dA\ B + A\ dB$$

We now multiply by $\overline{C}^H$ and take the real trace:

$$ds = \Re\left(\operatorname{tr}(\overline{C}^H dC)\right) = \Re\left(\operatorname{tr}(\overline{C}^H dA\ B)\right) + \Re\left(\operatorname{tr}(\overline{C}^H A\ dB)\right)$$

We use the `trace` identities to manipulate the $dA$ expression:

$$ds = \Re\left(\operatorname{tr}(B \overline{C}^H dA)\right) + \Re\left(\operatorname{tr}(\overline{C}^H A\ dB)\right)$$

That's it!

> The expression is in the desired form to solve for the adjoints:

Reviewer: Perhaps hold the reader's hand a little here by saying something like that you're matching terms in the above equation.

$$\overline{A} = (B \overline{C}^H)^H = \overline{C} B^H$$
$$\overline{B} = (\overline{C}^H A)^H = A^H \overline{C}$$

> Using our notation conventions, we would implement the `rrule` as

Reviewer: Consider `Using ChainRules' notation`, or something like that to distinguish from the notation in this doc.

```julia

> function rrule(::typeof(*), A::Matrix, B::Matrix)

Reviewer: ```suggestion
function rrule(::typeof(*), A::Matrix{Float64}, B::Matrix{Float64})
```

    function times_pullback(ΔC)
        ∂A = @thunk(ΔC * B')
        ∂B = @thunk(A' * ΔC)
        return (NO_FIELDS, ∂A, ∂B)
    end
    return A * B, times_pullback
end
```

### Matrix inversion

$$C A = A C = I$$

Again, we derived the differential identity as

> $$dC = -C dA C$$

Reviewer: inv again :)

Multiplying by $\overline{C}^H$ and taking the trace,

$$ds = \Re\left(\operatorname{tr}(\overline{C}^H dC)\right) = \Re\left(\operatorname{tr}(-\overline{C}^H C dA C)\right)$$

Applying the trace identities to manipulate into the desired form,

$$ds = \Re\left(\operatorname{tr}(-C \overline{C}^H C dA)\right),$$

we can now solve for $\overline{A}$:

$$\overline{A} = (-C \overline{C}^H C)^H = -C^H \overline{C} C^H$$

We can implement the resulting `rrule` as:

```julia

> function rrule(::typeof(inv), A::Matrix)

Reviewer: ```suggestion
function rrule(::typeof(inv), A::Matrix{Float64})
```

$$C_{i1k} = \sum_{j} |A_{ijk}|^2 = \sum_{j}\Re(A_{ijk}^* A_{ijk})$$

The differential identity is

$$dC_{i1k} = \sum_j \Re(dA_{ijk}^* A_{ijk} + A_{ijk}^* dA_{ijk}) \\
= \sum_j \Re((A_{ijk}^* dA_{ijk})^* + A_{ijk}^* dA_{ijk})\\
= \sum_j 2 \Re(A_{ijk}^* dA_{ijk}),$$

where in the last step we have used the identity that the sum of its number and its conjugate is twice the real part of the number.

We can then derive the reverse-mode rule.
The array form of the desired identity will be

$$ds = \Re \left(\sum_{ik} \overline{C}_{i1k}^* dC_{i1k}\right) = \Re \left(\sum_{ijk} \overline{A}_{ijk}^* dA_{ijk} \right)$$

We can plug in the differential identity into the middle expression to get

$$ds = \Re \left(\sum_{ijk} 2 \overline{C}_{i1k}^* \Re(A_{ijk}^* dA_{ijk})\right) = \Re \left(\sum_{ijk} 2 \Re(\overline{C}_{i1k}) A_{ijk}^* dA_{ijk}\right).$$

We can now solve for $\overline{A}$:

$$\overline{A}_{ijk} = (2 \Re(\overline{C}_{i1k}) A_{ijk}^*)^* = A_{ijk} \cdot 2\Re(\overline{C}_{i1k})$$

Because none of this derivation really depended on the index (or indices), we can easily implement the `rrule` more generically using broadcasting:

```julia

> function rrule(::typeof(sum), ::typeof(abs2), A::Array; dims = :)

Reviewer: ```suggestion
function rrule(::typeof(sum), ::typeof(abs2), A::Array{ComplexF64}; dims = :)
```

Author: These rules should work for any `Real` and `Complex`, so any objection to `Union{Real,Complex}`?

# Deriving Array Rules

> When the inputs and outputs of the functions are arrays (potentially with scalars), a modified version of Giles' method is a quick way of deriving [`frule`](@ref)s and [`rrule`](@ref)s.

Reviewer 4: Usually the second `s` is added if it would be pronouned (which i think it would be for "Giles's method"

```suggestion
When the inputs and outputs of the functions are arrays (potentially with scalars), a modified version of Giles's method is a quick way of deriving [`frule`](@ref)s and [`rrule`](@ref)s.
```

> Giles' method is succinctly explained in [^Giles2008] and its extended work [^Giles2008ext], but we will generalize it to handle arrays with both real and complex entries and arrays of arbitrary dimensions.

Reviewer 4: ```suggestion
Giles's method is succinctly explained in [^Giles2008] and its extended work [^Giles2008ext], but we will generalize it to handle arrays with both real and complex entries and arrays of arbitrary dimensions.
```

          \overline{C}^H \frac{\partial f}{\partial A} dA +
          \overline{C}^H \frac{\partial f}{\partial B} dB
      \right) \right) && \text{substitute } dC \text{ from } \eqref{cdiff}\\
   &= \operatorname{real}\left( \operatorname{tr}\left(
          \overline{C}^H \frac{\partial f}{\partial A} dA
      \right) \right) +
      \operatorname{real}\left( \operatorname{tr}\left(
          \overline{C}^H \frac{\partial f}{\partial B} dB
      \right) \right) && \text{expand using } \eqref{trexpand}
\end{align*}
```

By applying the same definition of $ds$ to the intermediates $A$ and $B$, we get

```math
ds = \operatorname{real}\left( \operatorname{tr}(\overline{A}^H  dA) \right) +
     \operatorname{real}\left( \operatorname{tr}(\overline{B}^H ~dB) \right)
```

Combining these two identities and solving for $\overline{A}$ and $\overline{B}$ gives us

```math
\begin{align*}
    \overline{A} &= (\overline{A}^H)^H
                  = \left( \overline{C}^H \frac{\partial f}{\partial A} \right)^H
                  = \left( \frac{\partial f}{\partial A} \right)^H \overline{C}\\
    \overline{B} &= \left( \frac{\partial f}{\partial B} \right)^H \overline{C}
\end{align*}
```

> Giles' method for deriving pullback functions is to first derive the differential identity (i.e. pushforward) using the above approach, then pre-multiply by $\overline{C}^H$, and take the real trace.

Reviewer 4: ```suggestion
Giles's method for deriving pullback functions is to first derive the differential identity (i.e. pushforward) using the above approach, then pre-multiply by $\overline{C}^H$, and take the real trace.
```

Reviewer 4: I wonder if this paragraph and the properties of the trace function should come before the example in which they are used? Or perhaps for this little section to have a sub-heading? It feels a bit lost amongst the text right now.

Subsequently, manipulate into this form and solve for the adjoint derivatives of the inputs.
Several properties of the trace function make this easier:

```math
\begin{align}
    \operatorname{tr}(A+B) &= \operatorname{tr}(A) + \operatorname{tr}(B) \label{trexpand}\\
    \operatorname{tr}(A^T) &= \operatorname{tr}(A) \nonumber\\
    \operatorname{tr}(A^H) &= \operatorname{conj}(\operatorname{tr}(A)) \nonumber\\
    \operatorname{tr}(AB) &= \operatorname{tr}(BA) \label{trperm}
\end{align}
```

!!! note

>     Our method is identical to Giles' approach, except we have replace the transpose with the conjugate transpose, and we have added the constraint that the inner product be real.

Reviewer 4: ```suggestion
    Our method is identical to Giles's method, except we have replace the transpose with the conjugate transpose, and we have added the constraint that the inner product be real.
```

we can now solve for $\overline{A}$:

$$\overline{A} = (-C \overline{C}^H C)^H = -C^H \overline{C} C^H$$

We can implement the resulting `rrule` as:

```julia
function rrule(::typeof(inv), A::Matrix{<:RealOrComplex})
    C = inv(A)
    function inv_pullback(ΔC)
        ∂A = -C' * ΔC * C'
        return (NO_FIELDS, ∂A)
    end
    return C, inv_pullback
end
```

## More examples

For more instructive examples of matrix and vector rules, see [^Giles2008ext] and the [LinearAlgebra rules in ChainRules](https://github.com/JuliaDiff/ChainRules.jl/tree/master/src/rulesets/LinearAlgebra).

## Generalizing to multidimensional arrays

For both forward- and reverse-mode rules for matrices, the first step was to write down the differential identities, which followed directly from the scalar differential identities.
This approach follows for arrays, but it's easier to work in component form.
Consider the following function

```julia
C = sum(abs2, A::Array{<:RealOrComplex,3}; dims=2)::Array{<:Real,3}
```

which we write as

$$C_{i1k} = \sum_{j} |A_{ijk}|^2 = \sum_{j} \operatorname{real}(\operatorname{conj}(A_{ijk}) A_{ijk})$$

The differential identity is

```math
\begin{align*}

>     dC_{i1k} &= \sum_j \operatorname{real}( \operatorname{conj}(dA_{ijk})~ A_{ijk} +

Reviewer 4: ```suggestion
    dC_{i1k} &= \sum_j \operatorname{real}\left( \operatorname{conj}(dA_{ijk})~ A_{ijk} +
```

>                             \operatorname{conj}(A_{ijk}) ~dA_{ijk} ) \\

Reviewer 4: ```suggestion
                            \operatorname{conj}(A_{ijk}) ~dA_{ijk} \right) \\
```

>              &= \sum_j \operatorname{real}(

Reviewer 4: ```suggestion
             &= \sum_j \operatorname{real}\left(
```

                    \operatorname{conj}\left(
                        \operatorname{conj}(A_{ijk}) ~dA_{ijk}
                    \right) +
                        \operatorname{conj}(A_{ijk}) ~dA_{ijk}

>                 )\\

Reviewer 4: ```suggestion
                \right\\
```

>              &= \sum_j 2 \operatorname{real}( \operatorname{conj}(A_{ijk}) ~dA_{ijk} )

Reviewer 4: ```suggestion
             &= \sum_j 2 \operatorname{real}\left( \operatorname{conj}(A_{ijk}) ~dA_{ijk} \right)
```

\end{align*}
```

where in the last step we have used the fact that for all real $a$ and $b$,

$$(a + i b) + \operatorname{conj}(a + i b) = (a + i b) + (a - i b) = 2 a = 2 \operatorname{real} (a + i b).$$

The `frule` can be implemented generally as

```julia
function frule(
    (_, _, ΔA),
    ::typeof(sum), ::typeof(abs2), A::Array{<:RealOrComplex};

>     dims = :,

Reviewer 4: should `dims=2`? since the primal function [shown above](https://github.com/JuliaDiff/ChainRulesCore.jl/pull/191/files#diff-04495d68483dba4eac9bf4246b9c62d0R342) has `dims=2`?

```suggestion
    dims=2,
```

)

>     C = sum(abs2, A; dims = dims)

Reviewer 4: ```suggestion
    C = sum(abs2, A; dims=dims)
```

Author: I ran the equations through JuliaFormatter, so I'm happy with the current spacing.

>     ∂C = sum(2 .* real.(conj.(A) .* ΔA); dims = dims)

Reviewer 4: ```suggestion
    ∂C = sum(2 .* real.(conj.(A) .* ΔA); dims=dims)
```

    return (C, ∂C)
end
```

We can now derive the reverse-mode rule.
The array form of the desired identity will be

```math
ds = \operatorname{real} \left( \sum_{ik}  \operatorname{conj}(\overline{C}_{i1k}) ~dC_{i1k} \right)
   = \operatorname{real} \left( \sum_{ijk} \operatorname{conj}(\overline{A}_{ijk}) ~dA_{ijk} \right)
```

We plug the differential identity into the middle expression to get

```math
\begin{align*}
    ds &= \operatorname{real} \left(\sum_{ijk}
                  \operatorname{conj}(\overline{C}_{i1k})

>                   2 \operatorname{real}(\operatorname{conj}(A_{ijk}) ~dA_{ijk})

Reviewer 4: ```suggestion
                  2 \operatorname{real}\left( \operatorname{conj}(A_{ijk}) ~dA_{ijk} \right)
```

              \right) \\
       &= \operatorname{real} \left( \sum_{ijk}
              2 \operatorname{real}(\overline{C}_{i1k})
              \operatorname{conj}(A_{ijk}) ~dA_{ijk}
          \right).
\end{align*}
```

We now solve for $\overline{A}$:

```math

> \overline{A}_{ijk} = \operatorname{conj}(

Reviewer 4: ```suggestion
\overline{A}_{ijk} = \operatorname{conj}\left(
```

                         2 \operatorname{real}( \overline{C}_{i1k} )
                         \operatorname{conj}(A_{ijk})

>                      )

Reviewer 4: ```suggestion
                     \right)
```

                   = 2\operatorname{real}( \overline{C}_{i1k} ) A_{ijk}
```

Because none of this derivation really depended on the index (or indices), we can easily implement the `rrule` more generically using broadcasting:

```julia

> function rrule(::typeof(sum), ::typeof(abs2), A::Array{<:RealOrComplex}; dims = :)

Reviewer 4: same question re `dims`?

Author: In this case I wanted to derive a generic rule but writing that out in math notation is horrific, so I instead derived for an arbitrary choice of indices and then stated that it's generic. I added some more text (hopefully) making that clearer.

    function sum_abs2_pullback(ΔC)
        ∂abs2 = DoesNotExist()
        ∂A = @thunk(2 .* real.(ΔC) .* A)
        return (NO_FIELDS, ∂abs2, ∂A)
    end

>     return sum(abs2, A; dims = dims), sum_abs2_pullback

Reviewer 4: ```suggestion
    return sum(abs2, A; dims=dims), sum_abs2_pullback
```

       && \text{substitute } A^{-1} = \Omega \nonumber\\
\dot{\Omega}
   &= -\Omega \dot{A} \Omega
       && \text{solve for } \dot{\Omega} \label{invdiff}
\end{align}
```

We write the `frule` as

```julia
function frule((_, ΔA), ::typeof(inv), A::Matrix{<:RealOrComplex})
    Ω = inv(A)
    ∂Ω = -Ω * ΔA * Ω
    return (Ω, ∂Ω)
end
```

### Other useful identities

These identities are particularly useful:

```math
\begin{align*}
\frac{d}{dt} \left( \operatorname{real}(A) \right) &= \operatorname{real}(\dot{A})\\
\frac{d}{dt} \left( \operatorname{conj}(A) \right) &= \operatorname{conj}(\dot{A})\\
\frac{d}{dt} \left( A^T \right) &= \dot{A}^T\\
\frac{d}{dt} \left( A^H \right) &= \dot{A}^H\\
\frac{d}{dt} \left( \sum_{j}  A_{i \ldots j \ldots k} \right) &=
    \sum_{j} \dot{A}_{i \ldots j \ldots k},
\end{align*}
```

where $\cdot^H = \operatorname{conj}(\cdot^T)$ is the conjugate transpose (the `adjoint` function).

## Reverse-mode rules

### Approach

Reverse-mode rules are a little less intuitive, but we can re-use our pushforwards to simplify their derivation.
Recall our program:

$$t \mapsto (\ldots, X_m, \ldots) \mapsto \Omega \mapsto s,$$

> At any step in the program, if we have intermediates $X_m$, we can write down the derivative $\frac{ds}{dt}$ in terms of the tangents $\dot{X}_m = \frac{dX_m}{dt}$ and adjoints $\overline{X}_m = \frac{\partial s}{\partial X_m}$

Reviewer 2: Should `$\overline{X}_m$` be superscript, or do you think that would be too confusing for the targeted audience?

Author: I considered it, but yeah, I think it would be too confusing. (this tutorial really should be using Einstein notation, which would really simplify the equations, but I also ruled that out for the same reason).

Reviewer 2: Yeah, seems reasonable. Probably also hard to keep this consistent throughout the whole tutorial. I don't think we can use Einstein notation here anyways, since sometimes twice-occuring indices are not summed over.

           \right)
       \right)
       \operatorname{tr}(A^{-1} \dot{A})
   \right)
       && \text{substitute } b \text{ from } \eqref{logabsdet_b} \\
&= \operatorname{real}\left( \operatorname{tr} \left(
       \left(
           \operatorname{real} \left( \overline{l} \right) +
           i \operatorname{imag} \left(
               \operatorname{conj} \left( \overline{s} \right) s
           \right)
       \right)
       A^{-1} \dot{A}
   \right) \right)
       && \text{bring scalar within } \operatorname{tr} \\
&= \operatorname{real}\left( \operatorname{tr} \left( \overline{A}^H \dot{A} \right) \right)
       && \text{right-hand side of } \eqref{pbidentmat}\\
\end{align*}
```

Now we solve for $\overline{A}$:

```math
\begin{align*}
\overline{A} &= \left( \left(
    \operatorname{real} \left( \overline{l} \right) +
    i \operatorname{imag} \left( \operatorname{conj} \left( \overline{s} \right) s \right)
\right) A^{-1} \right)^H\\
&= \left(
    \operatorname{real} \left( \overline{l} \right) +
    i \operatorname{imag} \left( \operatorname{conj} \left( s \right) \overline{s} \right)
\right) A^{-H}
\end{align*}
```

> !!! note

Reviewer 4: tiny things, but i wonder is this note can be moved to after the `rrule` code below? Since it discusses both pushforwards and pullbacks, it kinda breaks the flow of the section on Reverse-mode rule with where it is now

Author: Good point, thanks. Done!
